chips/stm32f429zi/src/ethernet/ethernet.rs

```diff
@@ -0,0 +1,2288 @@
+        match self.mac_registers.maca2hr.read(MACA2HR::AE) {
+            0 => false,
+            _ => true,
+        }
+    }
+
+    fn is_mac_address3_enabled(&self) -> bool {
+        match self.mac_registers.maca3hr.read(MACA3HR::AE) {
+            0 => false,
+            _ => true,
+        }
+    }
+
+    /* === DMA methods === */
+    fn reset_dma(&self) -> Result<(), ErrorCode> {
+        self.dma_registers.dmabmr.modify(DMABMR::SR::SET);
+
+        for _ in 0..100 {
+            if !self.dma_registers.dmabmr.is_set(DMABMR::SR) {
+                return Ok(());
+            }
+        }
+
+        Err(ErrorCode::FAIL)
+    }
+
+    fn dma_transmit_poll_demand(&self) {
+        self.dma_registers.dmatpdr.set(1);
+    }
+
+    fn dma_receive_poll_demand(&self) {
+        self.dma_registers.dmarpdr.set(1);
+    }
+
+    fn set_transmit_descriptor_list_address(&self, address: u32) -> Result<(), ErrorCode> {
+        if self.is_dma_transmission_enabled() == true {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.dma_registers.dmatdlar.set(address);
+
+        Ok(())
+    }
+
+    fn set_receive_descriptor_list_address(&self, address: u32) -> Result<(), ErrorCode> {
+        if self.is_dma_reception_enabled() == true {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.dma_registers.dmardlar.set(address);
+
+        Ok(())
+    }
+
+    fn get_transmit_descriptor_list_address(&self) -> u32 {
+        self.dma_registers.dmatdlar.get()
+    }
+
+    fn get_receive_descriptor_list_address(&self) -> u32 {
+        self.dma_registers.dmardlar.get()
+    }
+
+    fn get_transmit_process_state(&self) -> DmaTransmitProcessState {
+        match self.dma_registers.dmasr.read(DMASR::TPS) {
+            0b000 => DmaTransmitProcessState::Stopped,
+            0b001 => DmaTransmitProcessState::FetchingTransmitDescriptor,
+            0b010 => DmaTransmitProcessState::WaitingForStatus,
+            0b011 => DmaTransmitProcessState::ReadingData,
+            0b110 => DmaTransmitProcessState::Suspended,
+            _ => DmaTransmitProcessState::ClosingTransmitDescriptor,
+        }
+    }
+
+    fn get_receive_process_state(&self) -> DmaReceiveProcessState {
+        match self.dma_registers.dmasr.read(DMASR::RPS) {
+            0b000 => DmaReceiveProcessState::Stopped,
+            0b001 => DmaReceiveProcessState::FetchingReceiveDescriptor,
+            0b011 => DmaReceiveProcessState::WaitingForReceivePacket,
+            0b100 => DmaReceiveProcessState::Suspended,
+            0b101 => DmaReceiveProcessState::ClosingReceiveDescriptor,
+            _ => DmaReceiveProcessState::TransferringReceivePacketDataToHostMemory,
+        }
+    }
+
+    fn did_normal_interrupt_occur(&self) -> bool {
+        self.dma_registers.dmasr.is_set(DMASR::NIS)
+    }
+
+    // TODO: Am I allowed to clear this bit?
+    //fn clear_dma_normal_interrupt(&self) {
+    //self.dma_registers.dmasr.modify(DMASR::NIS::SET);
+    //}
```

Comment:
We should probably resolve these comments and add these functions or remove them appropriately. More generally, I think it's perfectly fine for a driver to not expose an exhaustive set of hardware capabilities, especially as accessor methods. Such methods typically suggest that some functionality is either required, or at least supported in a driver. Not implementing them can be interpreted as "we don't know whether touching this bit is safe, so we don't provide a method for it".

Reply:
> we don't know whether touching this bit is safe, so we don't provide a method for it

That's exactly the issue. The general description of DMA interrupts (page 1189):

> When all the enabled interrupts within a group are cleared, the corresponding summary bit is cleared.

The page describing registers (1226):

> The normal interrupt summary bit value is the logical OR of the following when the corresponding interrupt bits are enabled in the ETH_DMAIER register

So I far I understand that the summary bit is automatically cleared. However, the same page describing summary bits:

> This is a sticky bit and it must be cleared (by writing a 1 to this bit) each time a corresponding bit that causes NIS to be set is cleared
> This is a sticky bit and it must be cleared each time a corresponding bit that causes AIS to be
set is cleared

Do I miss something?